src/anndata/_core/aligned_mapping.py

```diff
@@ -272,20 +278,23 @@ def dim_names(self) -> pd.Index:
         return (self.parent.obs_names, self.parent.var_names)[self._axis]
 
 
+# TODO: vals can't be None
 class AxisArrays(AlignedActualMixin, AxisArraysBase):
     def __init__(
         self,
         parent: AnnData | Raw,
         axis: int,
-        vals: Mapping | AxisArraysBase | None = None,
+        vals: dict[str, V] | AxisArraysBase | None = None,
```

Comment:
Also, why is `Mapping` right everywhere?  Can there ever really be an immutable `vals`?

Reply:
I guess you're fix resolves that issue, actually.

Reply:
> I was trying to match this

that’s not a parameter, that’s a field. My point is that **usually**:
- function parameters should accept the most generic collections.abc interface (e.g. `Iterable` if possible, if not, then `Collection`, …)
- function return types should be concrete types
- class fields should have concrete types

> why is `Mapping` right everywhere?

because `_data: dict[...]` is a class member and `vals: Mapping[...]` is a parameter.